Web Inspector: add support for console snippets

dcrousso opened the pull request:
#### 23c03d7c6c3a03eca6396e8b563ec86f880a0dff
<pre>
Web Inspector: add support for console snippets
<a href="https://bugs.webkit.org/show_bug.cgi?id=245161">https://bugs.webkit.org/show_bug.cgi?id=245161</a>

Reviewed by Patrick Angle.

Console Snippets can be very useful for developers that frequently run the same JS over and over.

Previously, the developer would have to either repeatedly type the same thing, use the up arrow to
find the JS in the Console history, or even use the pasteboard (likely from another app/file).

With Console Snippets, the developer can rely on Web Inspector to organize and maintain the &quot;library&quot;
of JS that the developer might want to invoke, as well as having quick ways to actually do so.

* Source/WebInspectorUI/UserInterface/Models/ConsoleSnippet.js: Added.
(WI.ConsoleSnippet):
(WI.ConsoleSnippet.createDefaultWithTitle):
(WI.ConsoleSnippet.prototype.get title):
(WI.ConsoleSnippet.prototype.run):
(WI.ConsoleSnippet.fromJSON):
(WI.ConsoleSnippet.prototype.toJSON):
* Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
(WI.ConsoleManager):
(WI.ConsoleManager.prototype.get snippets): Added.
(WI.ConsoleManager.prototype.addSnippet): Added.
(WI.ConsoleManager.prototype.removeSnippet): Added.
(WI.ConsoleManager.prototype._handleSnippetContentChanged): Added.

* Source/WebInspectorUI/UserInterface/Views/ResourceIcons.css:
(:is(.anonymous-script-icon, .resource-icon.resource-type-script.console-snippet) .icon): Renamed from `.anonymous-script-icon .icon`.
(@media (prefers-color-scheme: dark) :is(.anonymous-script-icon, .resource-icon.resource-type-script.console-snippet) .icon): Renamed from `@media (prefers-color-scheme: dark) .anonymous-script-icon .icon`.
Force Console Snippets to use the JS clipping icon.

* Source/WebInspectorUI/UserInterface/Views/ConsoleSnippetTreeElement.js: Added.
(WI.ConsoleSnippetTreeElement):
(WI.ConsoleSnippetTreeElement.prototype.onattach):
(WI.ConsoleSnippetTreeElement.prototype.ondelete):
(WI.ConsoleSnippetTreeElement.prototype.onspace):
(WI.ConsoleSnippetTreeElement.prototype.canSelectOnMouseDown):
(WI.ConsoleSnippetTreeElement.prototype.updateStatus):
* Source/WebInspectorUI/UserInterface/Views/ConsoleSnippetTreeElement.css: Added.
(.tree-outline .item.console-snippet &gt; .status):
(.tree-outline .item.console-snippet &gt; .status &gt; img):
(.tree-outline .item.console-snippet:not(:hover) &gt; .status &gt; img):
(body:not(.window-inactive, .window-docked-inactive) .tree-outline:focus-within .item.console-snippet.selected:hover &gt; .status &gt; img):
(@media (prefers-color-scheme: dark) .tree-outline .item.console-snippet:hover &gt; .status &gt; img):
* Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js:
(WI.SourcesNavigationSidebarPanel):
(WI.SourcesNavigationSidebarPanel.prototype.closed):
(WI.SourcesNavigationSidebarPanel.prototype.createContentTreeOutline):
(WI.SourcesNavigationSidebarPanel.prototype.willDismissPopover):
(WI.SourcesNavigationSidebarPanel.prototype._willDismissConsoleSnippetPopover): Added.
(WI.SourcesNavigationSidebarPanel.prototype._compareTreeElements):
(WI.SourcesNavigationSidebarPanel.prototype._addConsoleSnippet): Added.
(WI.SourcesNavigationSidebarPanel.prototype._removeConsoleSnippet): Added.
(WI.SourcesNavigationSidebarPanel.prototype._handleCreateConsoleSnippetButtonClicked): Added.
(WI.SourcesNavigationSidebarPanel.prototype.async _populateCreateResourceContextMenu):
(WI.SourcesNavigationSidebarPanel.prototype._handleConsoleSnippetAdded): Added.
(WI.SourcesNavigationSidebarPanel.prototype._handleConsoleSnippetRemoved): Added.
* Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.css:
(.sidebar &gt; .panel.navigation.sources &gt; .content &gt; :matches(.pause-reason-container, .call-stack-container, .breakpoints-container, .local-overrides-container, .console-snippets-container)): Renamed from `.sidebar &gt; .panel.navigation.sources &gt; .content &gt; :matches(.pause-reason-container, .call-stack-container, .breakpoints-container, .local-overrides-container)`.
(@media (min-height: 650px) .sidebar &gt; .panel.navigation.sources &gt; .content &gt; :matches(.call-stack-container, .breakpoints-container, .resources-container, .local-overrides-container, .console-snippets-container)): Renamed from `@media (min-height: 650px) .sidebar &gt; .panel.navigation.sources &gt; .content &gt; :matches(.call-stack-container, .breakpoints-container, .resources-container, .local-overrides-container)`.
(@media (min-height: 650px) .sidebar &gt; .panel.navigation.sources &gt; .content &gt; .console-snippets-container): Added.
(@media (min-height: 650px) .sidebar &gt; .panel.navigation.sources &gt; .content &gt; .console-snippets-container &gt; .details-section): Added.
(@media (min-height: 650px) .sidebar &gt; .panel.navigation.sources &gt; .content &gt; .console-snippets-container &gt; .details-section &gt; .content): Added.
(.sidebar &gt; .panel.navigation.sources &gt; .content &gt; .local-overrides): Deleted.
Add a &quot;Console Snippets&quot; section to the navigation sidebar in the Sources Tab.

* Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
(WI.LogContentView):
(WI.LogContentView.prototype.get navigationItems):
(WI.LogContentView.prototype.willDismissPopover):
(WI.LogContentView.prototype._handleDrop):
(WI.LogContentView.prototype._handleSnippetsNavigationItemContextMenu): Added.
* Source/WebInspectorUI/UserInterface/Views/LogContentView.css:
(.navigation-bar &gt; .item.console-snippets &gt; img): Added.
(.console-user-command:not(:hover) &gt; .console-snippet-options): Added.
(.create-snippet-popover &gt; .label): Added.
(@media (prefers-color-scheme: dark) .navigation-bar &gt; .item.console-snippets &gt; img):
Add a snippets button navigation item to the Console for quick access when not in the Sources Tab.

* Source/WebInspectorUI/UserInterface/Views/OpenResourceDialog.js:
(WI.OpenResourceDialog.prototype.didDismissDialog):
(WI.OpenResourceDialog.prototype.didPresentDialog):
(WI.OpenResourceDialog.prototype._handleConsoleSnippetAdded): Added.
(WI.OpenResourceDialog.prototype._handleConsoleSnippetRemoved): Added.
* Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
(WI.SearchSidebarPanel.prototype.performSearch):
Support jumping to Console Snippets.

* Source/WebInspectorUI/UserInterface/Base/ObjectStore.js:
(WI.ObjectStore._open):
Bump the version for the new `console-snippets` store.

* Source/WebInspectorUI/UserInterface/Views/ButtonNavigationItem.js:
(WI.ButtonNavigationItem):
(WI.ButtonNavigationItem.prototype._update):
Wait to set the `title` until we know we know what `buttonStyle` we&apos;re using.

* Source/WebInspectorUI/UserInterface/Views/ConsoleCommandView.js:
(WI.ConsoleCommandView):
(WI.ConsoleCommandView.prototype.render):
Add a way for clients to add a `&quot;click&quot;` event listener after the command has `render()`.

* Source/WebInspectorUI/UserInterface/Views/ConsoleDrawer.css:
(.console-drawer &gt; .navigation-bar &gt; .item &gt; :is(.glyph, img), .console-drawer &gt; .navigation-bar &gt; .log-scope-bar &gt; li): Renamed from `.console-drawer &gt; .navigation-bar &gt; .item &gt; .glyph, .console-drawer &gt; .navigation-bar &gt; .log-scope-bar &gt; li`.
Allow `WI.ButtonNavigationItem.ImageType.IMG` to be clickable.

* Source/WebInspectorUI/UserInterface/Views/InputPopover.js:
(WI.InputPopover):
(WI.InputPopover.prototype.get identifier): Added.
(WI.InputPopover.prototype.get value):
(WI.InputPopover.prototype.show):
(WI.InputPopover.prototype.get result): Deleted.
It should be possible to get the `value` even if the developer has not hit enter/return.

* Source/WebInspectorUI/UserInterface/Views/ScriptContentView.js:
(WI.ScriptContentView.prototype._handleTextEditorContentDidChange):
Debounce how often we save the current contents after modifications (to prevent swamping IndexedDB).

* Source/WebInspectorUI/UserInterface/Base/Utilities.js:
(Set.prototype.some): Added.
Add utility function that behaves just like `Array.prototype.some`.

* Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
(WI.JavaScriptLogViewController):
(WI.JavaScriptLogViewController.prototype.startNewSession):
(WI.JavaScriptLogViewController.prototype.appendImmediateExecutionWithResult):
(WI.JavaScriptLogViewController.prototype.consolePromptTextCommitted):
(WI.JavaScriptLogViewController.prototype._appendConsoleMessageView):
* Source/WebInspectorUI/UserInterface/Controllers/DOMManager.js:
(WI.DOMManager.prototype.inspectNodeObject.nodeAvailable):
* Source/WebInspectorUI/UserInterface/Views/AnimationContentView.js:
(WI.AnimationContentView.prototype._populateAnimationTargetButtonContextMenu):
* Source/WebInspectorUI/UserInterface/Views/CanvasContentView.js:
(WI.CanvasContentView.prototype._populateCanvasElementButtonContextMenu):
* Source/WebInspectorUI/UserInterface/Views/CanvasTreeElement.js:
(WI.CanvasTreeElement.prototype.populateContextMenu):
* Source/WebInspectorUI/UserInterface/Views/ContextMenuUtilities.js:
* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotInstanceDataGridNode.js:
(WI.HeapSnapshotInstanceDataGridNode.logHeapSnapshotNode.node.shortestGCRootPath.):
(WI.HeapSnapshotInstanceDataGridNode.logHeapSnapshotNode):
* Source/WebInspectorUI/UserInterface/Views/ObjectPreviewView.js:
(WI.ObjectPreviewView.prototype._contextMenuHandler):
* Source/WebInspectorUI/UserInterface/Views/ObjectTreeBaseTreeElement.js:
(WI.ObjectTreeBaseTreeElement.prototype._logSymbolProperty):
(WI.ObjectTreeBaseTreeElement.prototype._logValue):
* Source/WebInspectorUI/UserInterface/Views/QuickConsole.js:
(WI.QuickConsole.prototype._handleDrop):
* Source/WebInspectorUI/UserInterface/Views/WebSocketDataGridNode.js:
(WI.WebSocketDataGridNode.prototype.appendContextMenuItems):
* Source/WebInspectorUI/UserInterface/Views/WebSocketResourceTreeElement.js:
(WI.WebSocketResourceTreeElement.prototype.populateContextMenu):
Drive-by: Rework parameters of `appendImmediateExecutionWithResult` to use an `options = {}`.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Images/Start.svg: Added.
* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Test.html:

Canonical link: <a href="https://commits.webkit.org/254636@main">https://commits.webkit.org/254636@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5041e0d00ce1c6b3283156b870f6aecc9deac4df

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89739 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34290 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/20451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/99084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/155900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32784 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/82108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/93434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95386 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/26053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/76584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/82108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/80930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/20451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/82108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30536 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/20451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/30285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/20451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33736 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/76584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1372 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32446 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/20451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
```diff
@@ -39,9 +39,28 @@ WI.ConsoleManager = class ConsoleManager extends WI.Object
+        WI.Target.registerInitializationPromise((async () => {
+            let serializedSnippets = await WI.objectStores.consoleSnippets.getAll();
+
+            this._restoringSnippets = true;
```
Should we declare this variable in the initializer itself as well to avoid mutating the shape of `ConsoleManager` in when doing this deferred work?

patrickangle commented on Source/WebInspectorUI/UserInterface/Test.html:
```diff
@@ -149,6 +149,7 @@
     <script src="Models/CollectionTypes.js"></script>
     <script src="Models/Color.js"></script>
     <script src="Models/ConsoleCommandResultMessage.js"></script>
+    <script src="Models/ConsoleSnippet.js"></script>
```
Should we have a test or two to test basic functionality, like making sure the default console snippet can run and returns the expected result?

dcrousso replied:
Hmm.

On the one hand, yes it would be good to have tests.

On the other hand, there's really nothing special about the functionality of this patch. It's built on top of existing stuff (e.g. `WI.RuntimeManager.prototype.evaluateInInspectedWindow`) and is really just about storing (BTW we don't support `WI.ObjectStore` inside tests to avoid any state across tests) and displaying the content.  I intentionally wanted this to be as similar to "copypaste the same text into the Console" as possible, just with a bit less "hassle" (i.e. a "Run" button to do that instead).  Also the default `WI.ConsoleSnippet` is currently just a big comment explaining what Console Snippets are with an expression for `"Hello World!"`, which I could totally see changing in the future if we find something better as the default.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ButtonNavigationItem.js:
```diff
@@ -221,16 +218,19 @@ WI.ButtonNavigationItem = class ButtonNavigationItem extends WI.NavigationItem
             }
 
             case WI.ButtonNavigationItem.ImageType.IMG: {
                 let img = this.element.appendChild(document.createElement("img"));
```
Should creating this element be inside the `if (this._image)`? For the SVG case we only create and append the element if we have `this._image`.

dcrousso replied:
This patch actually specifically wants a `<img>` even when `this._image` is not specified so that the `content` can be set in CSS.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/ButtonNavigationItem.js:
```diff
@@ -240,6 +240,9 @@ WI.ButtonNavigationItem = class ButtonNavigationItem extends WI.NavigationItem
         }
+
+        if (!this.tooltip && this._buttonStyle === WI.ButtonNavigationItem.Style.Image)
+            this.tooltip ||= this._label;
```
Either the `!this.tooltip` check or the `||=` is redundant.

patrickangle commented on Source/WebInspectorUI/UserInterface/Images/Start.svg:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0"?>
```
Should we just rename `AuditStart.svg` to `Start.svg` so that we use a consistent "start" arrow for running things?

dcrousso replied:
There is actually a minor difference between the two.  `AuditStart.svg` (and all the other `Audit*.svg`) are rounded, whereas `Start.svg` (and `Stop.svg`, `Pause.svg`, etc.) are all pointed.  This was an intentional design choice back when the Audit Tab was first created to better give a sense of familiarity with `xctest`.  Perhaps we could revisit that decision in the future.

dcrousso committed:
Web Inspector: add support for console snippets
https://bugs.webkit.org/show_bug.cgi?id=245161

Reviewed by Patrick Angle.

Console Snippets can be very useful for developers that frequently run the same JS over and over.

Previously, the developer would have to either repeatedly type the same thing, use the up arrow to
find the JS in the Console history, or even use the pasteboard (likely from another app/file).

With Console Snippets, the developer can rely on Web Inspector to organize and maintain the "library"
of JS that the developer might want to invoke, as well as having quick ways to actually do so.

* Source/WebInspectorUI/UserInterface/Models/ConsoleSnippet.js: Added.
(WI.ConsoleSnippet):
(WI.ConsoleSnippet.createDefaultWithTitle):
(WI.ConsoleSnippet.prototype.get title):
(WI.ConsoleSnippet.prototype.run):
(WI.ConsoleSnippet.fromJSON):
(WI.ConsoleSnippet.prototype.toJSON):
* Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
(WI.ConsoleManager):
(WI.ConsoleManager.prototype.get snippets): Added.
(WI.ConsoleManager.prototype.addSnippet): Added.
(WI.ConsoleManager.prototype.removeSnippet): Added.
(WI.ConsoleManager.prototype._handleSnippetContentChanged): Added.

* Source/WebInspectorUI/UserInterface/Views/ResourceIcons.css:
(:is(.anonymous-script-icon, .resource-icon.resource-type-script.console-snippet) .icon): Renamed from `.anonymous-script-icon .icon`.
(@media (prefers-color-scheme: dark) :is(.anonymous-script-icon, .resource-icon.resource-type-script.console-snippet) .icon): Renamed from `@media (prefers-color-scheme: dark) .anonymous-script-icon .icon`.
Force Console Snippets to use the JS clipping icon.

* Source/WebInspectorUI/UserInterface/Views/ConsoleSnippetTreeElement.js: Added.
(WI.ConsoleSnippetTreeElement):
(WI.ConsoleSnippetTreeElement.prototype.onattach):
(WI.ConsoleSnippetTreeElement.prototype.ondelete):
(WI.ConsoleSnippetTreeElement.prototype.onspace):
(WI.ConsoleSnippetTreeElement.prototype.canSelectOnMouseDown):
(WI.ConsoleSnippetTreeElement.prototype.updateStatus):
* Source/WebInspectorUI/UserInterface/Views/ConsoleSnippetTreeElement.css: Added.
(.tree-outline .item.console-snippet > .status):
(.tree-outline .item.console-snippet > .status > img):
(.tree-outline .item.console-snippet:not(:hover) > .status > img):
(body:not(.window-inactive, .window-docked-inactive) .tree-outline:focus-within .item.console-snippet.selected:hover > .status > img):
(@media (prefers-color-scheme: dark) .tree-outline .item.console-snippet:hover > .status > img):
* Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js:
(WI.SourcesNavigationSidebarPanel):
(WI.SourcesNavigationSidebarPanel.prototype.closed):
(WI.SourcesNavigationSidebarPanel.prototype.createContentTreeOutline):
(WI.SourcesNavigationSidebarPanel.prototype.willDismissPopover):
(WI.SourcesNavigationSidebarPanel.prototype._willDismissConsoleSnippetPopover): Added.
(WI.SourcesNavigationSidebarPanel.prototype._compareTreeElements):
(WI.SourcesNavigationSidebarPanel.prototype._addConsoleSnippet): Added.
(WI.SourcesNavigationSidebarPanel.prototype._removeConsoleSnippet): Added.
(WI.SourcesNavigationSidebarPanel.prototype._handleCreateConsoleSnippetButtonClicked): Added.
(WI.SourcesNavigationSidebarPanel.prototype.async _populateCreateResourceContextMenu):
(WI.SourcesNavigationSidebarPanel.prototype._handleConsoleSnippetAdded): Added.
(WI.SourcesNavigationSidebarPanel.prototype._handleConsoleSnippetRemoved): Added.
* Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.css:
(.sidebar > .panel.navigation.sources > .content > :matches(.pause-reason-container, .call-stack-container, .breakpoints-container, .local-overrides-container, .console-snippets-container)): Renamed from `.sidebar > .panel.navigation.sources > .content > :matches(.pause-reason-container, .call-stack-container, .breakpoints-container, .local-overrides-container)`.
(@media (min-height: 650px) .sidebar > .panel.navigation.sources > .content > :matches(.call-stack-container, .breakpoints-container, .resources-container, .local-overrides-container, .console-snippets-container)): Renamed from `@media (min-height: 650px) .sidebar > .panel.navigation.sources > .content > :matches(.call-stack-container, .breakpoints-container, .resources-container, .local-overrides-container)`.
(@media (min-height: 650px) .sidebar > .panel.navigation.sources > .content > .console-snippets-container): Added.
(@media (min-height: 650px) .sidebar > .panel.navigation.sources > .content > .console-snippets-container > .details-section): Added.
(@media (min-height: 650px) .sidebar > .panel.navigation.sources > .content > .console-snippets-container > .details-section > .content): Added.
(.sidebar > .panel.navigation.sources > .content > .local-overrides): Deleted.
Add a "Console Snippets" section to the navigation sidebar in the Sources Tab.

* Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
(WI.LogContentView):
(WI.LogContentView.prototype.get navigationItems):
(WI.LogContentView.prototype.willDismissPopover):
(WI.LogContentView.prototype._handleDrop):
(WI.LogContentView.prototype._handleSnippetsNavigationItemContextMenu): Added.
* Source/WebInspectorUI/UserInterface/Views/LogContentView.css:
(.navigation-bar > .item.console-snippets > img): Added.
(.console-user-command:not(:hover) > .console-snippet-options): Added.
(.create-snippet-popover > .label): Added.
(@media (prefers-color-scheme: dark) .navigation-bar > .item.console-snippets > img):
Add a snippets button navigation item to the Console for quick access when not in the Sources Tab.

* Source/WebInspectorUI/UserInterface/Views/OpenResourceDialog.js:
(WI.OpenResourceDialog.prototype.didDismissDialog):
(WI.OpenResourceDialog.prototype.didPresentDialog):
(WI.OpenResourceDialog.prototype._handleConsoleSnippetAdded): Added.
(WI.OpenResourceDialog.prototype._handleConsoleSnippetRemoved): Added.
* Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
(WI.SearchSidebarPanel.prototype.performSearch):
Support jumping to Console Snippets.

* Source/WebInspectorUI/UserInterface/Base/ObjectStore.js:
(WI.ObjectStore._open):
Bump the version for the new `console-snippets` store.

* Source/WebInspectorUI/UserInterface/Views/ButtonNavigationItem.js:
(WI.ButtonNavigationItem):
(WI.ButtonNavigationItem.prototype._update):
Wait to set the `title` until we know we know what `buttonStyle` we're using.

* Source/WebInspectorUI/UserInterface/Views/ConsoleCommandView.js:
(WI.ConsoleCommandView):
(WI.ConsoleCommandView.prototype.render):
Add a way for clients to add a `"click"` event listener after the command has `render()`.

* Source/WebInspectorUI/UserInterface/Views/ConsoleDrawer.css:
(.console-drawer > .navigation-bar > .item > :is(.glyph, img), .console-drawer > .navigation-bar > .log-scope-bar > li): Renamed from `.console-drawer > .navigation-bar > .item > .glyph, .console-drawer > .navigation-bar > .log-scope-bar > li`.
Allow `WI.ButtonNavigationItem.ImageType.IMG` to be clickable.

* Source/WebInspectorUI/UserInterface/Views/InputPopover.js:
(WI.InputPopover):
(WI.InputPopover.prototype.get identifier): Added.
(WI.InputPopover.prototype.get value):
(WI.InputPopover.prototype.show):
(WI.InputPopover.prototype.get result): Deleted.
It should be possible to get the `value` even if the developer has not hit enter/return.

* Source/WebInspectorUI/UserInterface/Views/ScriptContentView.js:
(WI.ScriptContentView.prototype._handleTextEditorContentDidChange):
Debounce how often we save the current contents after modifications (to prevent swamping IndexedDB).

* Source/WebInspectorUI/UserInterface/Base/Utilities.js:
(Set.prototype.some): Added.
Add utility function that behaves just like `Array.prototype.some`.

* Source/WebInspectorUI/UserInterface/Controllers/JavaScriptLogViewController.js:
(WI.JavaScriptLogViewController):
(WI.JavaScriptLogViewController.prototype.startNewSession):
(WI.JavaScriptLogViewController.prototype.appendImmediateExecutionWithResult):
(WI.JavaScriptLogViewController.prototype.consolePromptTextCommitted):
(WI.JavaScriptLogViewController.prototype._appendConsoleMessageView):
* Source/WebInspectorUI/UserInterface/Controllers/DOMManager.js:
(WI.DOMManager.prototype.inspectNodeObject.nodeAvailable):
* Source/WebInspectorUI/UserInterface/Views/AnimationContentView.js:
(WI.AnimationContentView.prototype._populateAnimationTargetButtonContextMenu):
* Source/WebInspectorUI/UserInterface/Views/CanvasContentView.js:
(WI.CanvasContentView.prototype._populateCanvasElementButtonContextMenu):
* Source/WebInspectorUI/UserInterface/Views/CanvasTreeElement.js:
(WI.CanvasTreeElement.prototype.populateContextMenu):
* Source/WebInspectorUI/UserInterface/Views/ContextMenuUtilities.js:
* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotInstanceDataGridNode.js:
(WI.HeapSnapshotInstanceDataGridNode.logHeapSnapshotNode.node.shortestGCRootPath.):
(WI.HeapSnapshotInstanceDataGridNode.logHeapSnapshotNode):
* Source/WebInspectorUI/UserInterface/Views/ObjectPreviewView.js:
(WI.ObjectPreviewView.prototype._contextMenuHandler):
* Source/WebInspectorUI/UserInterface/Views/ObjectTreeBaseTreeElement.js:
(WI.ObjectTreeBaseTreeElement.prototype._logSymbolProperty):
(WI.ObjectTreeBaseTreeElement.prototype._logValue):
* Source/WebInspectorUI/UserInterface/Views/QuickConsole.js:
(WI.QuickConsole.prototype._handleDrop):
* Source/WebInspectorUI/UserInterface/Views/WebSocketDataGridNode.js:
(WI.WebSocketDataGridNode.prototype.appendContextMenuItems):
* Source/WebInspectorUI/UserInterface/Views/WebSocketResourceTreeElement.js:
(WI.WebSocketResourceTreeElement.prototype.populateContextMenu):
Drive-by: Rework parameters of `appendImmediateExecutionWithResult` to use an `options = {}`.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Images/Start.svg: Added.
* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Test.html:

Canonical link: https://commits.webkit.org/254636@main

JosephPecoraro commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSnippetTreeElement.js:
```diff
@@ -0,0 +1,76 @@
+        return true;
+    }
+
+    onspace()
```
Should `onenter` do the same thing?

JosephPecoraro commented on Source/WebInspectorUI/UserInterface/Views/ConsoleSnippetTreeElement.js:
```diff
@@ -0,0 +1,76 @@
+        super.onattach();
+
+        this.status = document.createElement("img");
+        this.status.title = WI.UIString("Run");
```
Would be nice to have `Run` be a Context Menu item for this model object.

JosephPecoraro commented on Source/WebInspectorUI/UserInterface/Models/ConsoleSnippet.js:
```diff
@@ -0,0 +1,110 @@
+                    if (!WI.consoleManager.snippets.has(this))
+                        return;
+
+
```
Extra newline!

JosephPecoraro commented on Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js:
```diff
@@ -2176,6 +2293,19 @@ WI.SourcesNavigationSidebarPanel = class SourcesNavigationSidebarPanel extends W
             });
         }
 
+        contextMenu.appendItem(WI.UIString("Console Snippet..."), () => {
```
Other localized string above uses `\u2026` instead of `...`.

JosephPecoraro commented on Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
```diff
@@ -88,6 +109,33 @@ WI.ConsoleManager = class ConsoleManager extends WI.Object
+
+        this._snippets.delete(snippet);
+
+        if (!this._restoringSnippets)
```
Possible cleanup. This could probably be an assert. I don't think we'd ever expect a remove during restoration? `console.assert(!this._restoringSnippets)`. But it is nice to be symmetric with `addSnippet`.

JosephPecoraro replied:
I suppose someone could technically remove a snippet in response to `WI.ConsoleManager.Event.SnippetAdded` during restoration... But I like the assertion because we'd never expect that.

JosephPecoraro commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -926,6 +964,21 @@ WI.LogContentView = class LogContentView extends WI.ContentView
         }, !WI.settings.clearLogOnNavigate.value);
     }
 
+    _handleSnippetsNavigationItemContextMenu(contextMenu) {
```
Style: `{` on new line.

JosephPecoraro commented on Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
```diff
@@ -926,6 +964,21 @@ WI.LogContentView = class LogContentView extends WI.ContentView
+
+        contextMenu.appendSeparator();
+
+        contextMenu.appendItem(WI.UIString("Create Console Snippet..."), () => {
```
Another case for `\u2026` instead of actual dots.
